src/lib/raft/raft.c

```diff
@@ -206,10 +206,51 @@ raft_scores_str(const struct raft *raft)
 	return buf;
 }
 
+static inline bool
+raft_leader_is_seen(const struct raft *raft)
+{
+	bool seen = bit_test(&raft->leader_witness_map, raft->self);
+	assert(!seen || raft->leader != 0);
+	return seen;
+}
+
 /** Schedule broadcast of the complete Raft state to all the followers. */
 static void
 raft_schedule_broadcast(struct raft *raft);
 
+/**
+ * Update all leader-related fields on leader appearance and let others know the
+ * leader is seen.
+ */
+static inline void
+raft_set_leader(struct raft *raft, uint32_t leader)
+{
+	assert(raft->leader == 0);
+	assert(leader != 0);
+	raft->leader = leader;
+	bool changed = raft_notify_leader_seen(raft, true, raft->self);
+	assert(changed);
+	(void)changed;
+	raft->leader_last_seen = raft_ev_now(raft_loop());
+	/* Leader is seen. Tell other nodes about its appearance. */
+	raft_schedule_broadcast(raft);
+}
+
+/**
+ * Clear leader-related fields on leader loss or resign and let everyone know
+ * this node doesn't see the leader anymore.
+ */
+static inline void
+raft_clear_leader(struct raft *raft)
+{
+	if (raft->leader == raft->self)
+		raft->leader_last_seen = raft_ev_now(raft_loop());
```

Comment:
It's for the sake of `box.info.election.leader_idle`. While the node is a leader, its `leader_idle` is always zero. I'd like `leader_idle` to start counting time from the moment of its resign.

Reply:
Ok, I see now. Deserves a comment I think.

Reply:
Done:
```diff
diff --git a/src/lib/raft/raft.c b/src/lib/raft/raft.c
index ccd716e22..6a831a6e4 100644
--- a/src/lib/raft/raft.c
+++ b/src/lib/raft/raft.c
@@ -243,8 +243,13 @@ raft_set_leader(struct raft *raft, uint32_t leader)
 static inline void
 raft_clear_leader(struct raft *raft)
 {
-       if (raft->leader == raft->self)
+       if (raft->leader == raft->self) {
+               /*
+                * Update leader_last_seen when resigning so that leader_idle
+                * starts counting from zero after the resign.
+                */
                raft->leader_last_seen = raft_ev_now(raft_loop());
+       }
        raft->leader = 0;
        bool is_changed = raft_notify_is_leader_seen(raft, false, raft->self);
        if (is_changed)
```